### PR TITLE
Text 2 tree builder

### LINKED ISSUE #2028: Build the plain text to tree builder.
This object should be called the `HTMLTreeBuilder` which builds a tree from an HTML document.

We want to use a tried and true HTML parser. In my investigation [`parse5`](https://github.com/inikulin/parse5) is the HTML parser we want to use. There are essentially two ways to implement the tree builder:

1. Let `parse5` parse the complete document and then convert this HTML tree into our structured tree.
2. Implement a `TreeAdapter`: https://github.com/inikulin/parse5/blob/master/packages/parse5/docs/options/parser-options.md#optional-treeadapter. This prevents the building of an intermediate HTML tree before we move to our own tree.

Let's try option 2 first because it will be the most efficient, because we can skip building the intermediate HTML tree. If option 2 is not feasible for some reason we can fallback to option 1.

## Examples

Input:

```html
<section>
	<div>
		This <strong>sentence</strong> <strong class="className" data-extra="pieceOfData">needs to <em>be</em> read</strong> to have value as a sentence.
	</div>
</section>
```

Output:

```js
tree = {
	children: [
		new StructuredNode( {
			children: [
				new StructuredNode( {
					children: [
						new Paragraph( {
							"text": new Text( {
								"text": "This sentence needs to be read to have value as a sentence.",
								"formatting": [
									{
										tag: "strong",
										attributes: null,
										start: 5, // "This ".length
										end: 13, // "This sentence".length
									},
									{
										tag: "strong",
										attributes: {
											class: "className",
											"data-extra": "pieceOfData"
										},
										start: 14, // "This sentence ".length
										end: 30, // "This sentence needs to be read".length
									},
									{
										tag: "em",
										attributes: null,
										start: 23, // "This sentence needs to ".length
										end: 25, // "This sentence needs to be".length
									}
								]
							} )
						} )
					]
				} ),
			],
		} ),
	],
};
```

```html
<script type="text/javascript">
	document.onload = () => { console.log( "Page loaded!" ); };
</script>

<section>
	<h2>First section</h2>

	<ul>
		<li>First item</li>
		<li>Second item</li>
		<li>Third item</li>
	</ul>

	Paragraph

	<img src="https://yoast.com/image-path" alt="Awesome Image" />
</section>

<section class="secondSection">
	<h2>Second section</h2>

	<ol>
		<li>First item</li>
		<li>Second item</li>
		<li>Third item</li>
	</ol>

	First paragraph

	Second paragraph
</section>
```

```js
tree = {
	children: [
		new StructuredIrrelevant( {
			content: '<script type="text/javascript">\n\tdocument.onload = () => { console.log( "Page loaded!" ); };\n</script>',
		} ),
		new StructuredText( {
			children: [
				new Heading( {
					text: new Text( {
						text: "First section",						
					} ),
                                        level: 2,
					startHtml: "<h2>",
					endHtml: "</h2>",
				} ),

				new List( {
					type: "unordered",
					startHtml: "<ul>",
					endHtml: "</ul>",
					items: [
						new ListItem( {
							text: new Text( {
								text: "First item",
							} )
						} ),
						new ListItem( {
							text: new Text( {
								text: "Second item",
							} )
						} ),
						new ListItem( {
							text: new Text( {
								text: "Third item",
							} )
						} ),
					],
				} ),

				new Paragraph( {
					text: new Text( {
						text: "Paragraph",
					} ),
				} ),

				new Image( {
					content: '<img src="https://yoast.com/image-path" alt="Awesome Image" />',
					attributes: {
						src: "https://yoast.com/image-path",
						alt: "Awesome Image",
					},
				} ),
			],
			startHtml: "<section>",
			endHtml: "</section>",
		} ),
		new StructuredText( {
			children: [
				new Heading( {
					text: new Text( {
						text: "Second section",						
					} ),
                                        level: 2,
					startHtml: "<h2>",
					endHtml: "</h2>",
				} ),

				new List( {
					type: "ordered",
					startHtml: "<ul>",
					endHtml: "</ul>",
					items: [
						new ListItem( {
							text: new Text( {
								text: "First item",
							} )
						} ),
						new ListItem( {
							text: new Text( {
								text: "Second item",
							} )
						} ),
						new ListItem( {
							text: new Text( {
								text: "Third item",
							} )
						} ),
					],
				} ),

				new Paragraph( {
					text: new Text( {
						text: "First paragraph",
					} ),
				} ),

				new Paragraph( {
					text: new Text( {
						text: "Second paragraph",
					} ),
				} ),
			],
			startHtml: "<section>",
			endHtml: "</section>",
		} ),
	]
};
```

## Misc

Use the `parseFragment` method from `parse5`, because `parse` will include `html`, `head` and `body` nodes. We don't need those nodes.

Parent #1903.

### PR DESCRIPTION
**Pull request by nataliashitova**
## Summary

This PR can be summarized in the following changelog entry:

* _[not-user-facing]_ Adds functionality to build a structured tree representation of the source text to analyze.

### Note for testers.
This functionality is not yet used in production. In the (near) future the assessments will be (one-by-one) rewritten to use this tree representation, instead of the current string representation.

## Relevant technical choices:

* Adds the `parse5` library as a basic HTML parser.
* Implements a `TreeAdapter` interface to generate a tree to fit our needs.

## Test instructions

This PR can be tested by following these steps:

* Run `yarn test`, check if the tests pass and the coverage is 100%.
* Review the tests in `buildTreeSpec.js` and the added code used for building the tree. See if you can think of any (edge) cases that are not tested yet and add them as comments to this PR.

Fixes #2028

### CODE REVIEW

File: spec/tree/values/nodes/ParagraphSpec.js [hunk 1]
```diff
@@ -1,43 +1,30 @@
...
-			const startIndex = 0;
-			const endIndex = 30;
-			const tag = "p";
-
-			const paragraphElement = new Paragraph( textElement, startIndex, endIndex, tag );
-
-			expect( paragraphElement.textContainer ).toEqual( textElement );
-			expect( paragraphElement.startIndex ).toEqual( 0 );
-			expect( paragraphElement.endIndex ).toEqual( 30 );
+		test( "can make a new Paragraph tree node", () => {
```
Comment (atimmer) [inline spec/tree/values/nodes/ParagraphSpec.js]: Everywhere else we use `it`, we should adhere to that.

Also then this text needs to be changed to fit that. So `it( "makes a new Paragraph tree node"`. But I think `creates` sounds better here.

```suggestion
		it( "creates a new Paragraph tree node", () => {
```

File: spec/tree/builder/buildTreeSpec.js [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+		section.startIndex = 0;
+		section.endIndex = 62;
+		section.children = [ structuredNode ];
+
+		const expected = new StructuredNode( "root" );
+		expected.startIndex = 0;
+		expected.endIndex = 62;
+		expected.children = [ section ];
+
+		expect(  treeToStringifiedJSON( tree ) ).toEqual( JSON.stringify( expected ) );
```
Comment (atimmer) [inline spec/tree/builder/buildTreeSpec.js]:
```suggestion
		expect( treeToStringifiedJSON( tree ) ).toEqual( JSON.stringify( expected ) );
```

File: package.json [hunk 3]
```diff
@@ -64,6 +64,8 @@
     "jed": "^1.1.0",
     "lodash-es": "^4.17.10",
     "loglevel": "^1.6.1",
+    "parse5": "^5.1.0",
+    "parse5-htmlparser2-tree-adapter": "^5.1.0",
```
Comment (atimmer) [inline package.json]: Where is this used in our code?

File: src/tree/builder/TreeAdapter.js [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+
+	/**
+	 * Detaches a node from its parent.
+	 *
+	 * @param {Node} node	The node to detach from its parent.
+	 *
+	 * @returns {void}
+	 */
+	detachNode( node ) {
+		if ( node.parent ) {
```
Comment (igorschoester) [inline src/tree/builder/TreeAdapter.js]: It seems the else statement of this if is never reached in the coverage files.

File: src/tree/builder/htmlClasses.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+/**
+ * All HTML element tags that we treat as formatting.
+ * @type {string[]}
+ * @const
+ *
+ * @memberOf module:tree/builder
+ */
+const formattingElements = [ "a", "abbr", "audio", "b", "bdo", "br", "button", "canvas", "cite", "code", "command",
+	"data", "datalist", "dfn", "del", "em", "embed", "i", "iframe", "img", "input", "kbd", "keygen", "label", "mark",
+	"math", "meter", "noscript", "object", "output", "progress", "q", "ruby", "samp", "script", "select", "small",
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/htmlClasses.js]: I am using [this list on HTML phrasing content on Mozilla Developer Network](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#Phrasing_content).

I think having a structure like this is valid HTML according to MDN:
```html
<p>This is some script <script>console.log("logging something");</script></p>
```
So I think it should be added as formatting to the paragraph for completeness.

I think it would not hurt for me to add a `@see` JSDoc tag to point towards the MDN list I used, for documentation purposes.

File: spec/tree/builder/buildTreeSpec.js [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+
+		const expected = new StructuredNode( "root" );
+		expected.startIndex = 0;
+		expected.endIndex = 80;
+		expected.children = [ section ];
+
+		expect(  tree.toString() ).toEqual( expected.toString() );
+	} );
+
+	it( "can parse an irrelevant HTML-element and its contents into a StructuredIrrelevant node.", () => {
```
Comment (agnieszkaszuba) [inline spec/tree/builder/buildTreeSpec.js]:
```suggestion
	it( "can parse an irrelevant HTML element and its contents into a StructuredIrrelevant node.", () => {
```

File: src/tree/structure/nodes/StructuredIrrelevant.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+	 * Examples from HTML include content within `<script>` and `<style>` elements.
+	 *
+	 * @param {string} tag The element tag.
+	 *
+	 * @returns {void}
+	 */
+	constructor( tag ) {
+		super( "StructuredIrrelevant" );
+		/**
+		 * Type of node (e.g. "script", "code" etc.).
```
Comment (igorschoester) [inline src/tree/structure/nodes/StructuredIrrelevant.js]: `code` is not in the irrelevant list at this time.

File: src/tree/builder/TreeAdapter.js [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+ *
+ * @private
+ */
+class TreeAdapter {
+	// Creation of nodes and other tree elements.
+
+	/**
+	 * Creates a new element to be put in the tree.
+	 *
+	 * @param {string} tag									The HTML tag.
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Okay, replaced the tabs with spaces in the JSDoc comments 🙂.

File: src/tree/builder/TreeAdapter.js [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+	/**
+	 * Makes a new node to add to the tree, based on the given HTML-tag.
+	 *
+	 * @param {string} tag		The HTML-tag of the element to add.
+	 *
+	 * @returns {module:tree/structure.Node} The node to add to the tree.
+	 *
+	 * @private
+	 */
+	_parseNode( tag ) {
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Made it into a static function.

File: src/tree/builder/TreeAdapter.js [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+	 *
+	 * @returns {void}
+	 */
+	appendChild( parent, child ) {
+		// Do not do anything with irrelevant content.
+		if ( parent instanceof StructuredIrrelevant ) {
+			return;
+		}
+
+		if ( child instanceof StructuredIrrelevant &&
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: I added a comment! :)

File: src/tree/builder/TreeAdapter.js [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+	}
+
+	// Private utility methods.
+
+	/**
+	 * Finds the most recent ancestor (parent of parent of ... ) of this node that returns true
+	 * on the given predicate.
+	 *
+	 * @param {Node|FormattingElement} element  The node to find the ancestor of.
+	 * @param {predicate} predicate	             The predicate to check the ancestors on.
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: You're right, this was a remainder of a previous iteration of the function, changed it to `function`!

File: src/tree/builder/TreeAdapter.js [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+	 *
+	 * @returns {void}
+	 *
+	 * @private
+	 */
+	_addStructuredNodeText( node, text ) {
+		// Get the previous sibling of this node.
+		const prevChild = node.children[ node.children.length - 1 ];
+		// If the previous child is an implicit paragraph...
+		if ( prevChild && prevChild instanceof Paragraph && ! prevChild.isExplicit() ) {
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Added an example!

File: src/tree/builder/TreeAdapter.js [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+	 * Gets a parent's first child.
+	 *
+	 * @see https://en.wikipedia.org/wiki/Rumpelstiltskin
+	 *
+	 * @param {Node} node	The node to get its first child from.
+	 *
+	 * @returns {Node[]|null} The node's first child or null, if this node cannot get any children.
+	 */
+	getFirstChild( node ) {
+		return node.children[ 0 ] || null;
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Added a check if `node.children` exists, if not it returns `null`.

File: src/tree/builder/TreeAdapter.js [hunk 14]
```diff
@@ -0,0 +1,449 @@
...
+	 * @private
+	 */
+	static _parseAttributes( parse5attributes ) {
+		if ( parse5attributes ) {
+			return parse5attributes.reduce( ( attributes, attribute ) => {
+				attributes[ attribute.name ] = attribute.value;
+				return attributes;
+			}, {} );
+		}
+		return {};
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: I changed it, it now returns `null` instead of the empty object. In general: the Wiki on the tree structure is outdated, I will update the Wiki after the code has stabilized a bit, to avoid superfluous work.

I added an extra test for the `TreeAdapter` to check if the attributes are `null` when a formatting element with no attributes is added.

File: src/tree/builder/TreeAdapter.js [hunk 15]
```diff
@@ -0,0 +1,449 @@
...
+	 * @private
+	 */
+	_parseNode( tag ) {
+		if ( headings.includes( tag ) ) {
+			// Heading.
+			return new Heading( parseInt( tag[ 1 ], 10 ) );
+		} else if ( tag === "p" ) {
+			// Paragraph.
+			return new Paragraph( tag );
+		} else if ( tag === "ol" || tag === "ul" ) {
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Do you mean `dl` lists? 

They are somewhat different than ordered and unordered lists in that you must specify individual keys, using `dt` tags. Which the current code does not support yet.

(In the case of ordered and unordered lists, the keys can be deduced by the kind of list instead (e.g. "1, 2, 3, ..." in the case of ordered, bullet points in the case of unordered lists).)

Since this PR already adds huge chunk of new functionality, I would make an issue and add support for `dl` lists later on instead.

[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Made a new issue: #2094.

File: src/tree/builder/TreeAdapter.js [hunk 16]
```diff
@@ -0,0 +1,449 @@
...
+		} else if ( formattingElements.includes( tag ) ) {
+			// Formatting element.
+			const parsedAttributes = TreeAdapter._parseAttributes( attributes );
+			node = new FormattingElement( tag, parsedAttributes );
+		} else {
+			// Paragraphs, Headers, Lists, ListItems and other nodes.
+			node = this._parseNode( tag );
+		}
+
+		node.tagName = tag;
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Added a comment explaining why it is needed 🙂.

File: src/tree/builder/TreeAdapter.js [hunk 17]
```diff
@@ -0,0 +1,449 @@
...
+			return parse5attributes.reduce( ( attributes, attribute ) => {
+				attributes[ attribute.name ] = attribute.value;
+				return attributes;
+			}, {} );
+		}
+		return {};
+	}
+
+	/**
+	 * Creates a new empty document fragment (e.g. a part of an HTML document).
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: A document fragment is any HTML that is not a full-fledged HTML document.
Like:
```html
<h1>Header</h1>
<p>Paragraph</p>
```
Every tree has to have one, and only one, root node. `createDocumentFragment` is the function `parse5` uses to make one.

`root` is not part of the HTML specification, but a "type" of node of my own devising. It is a way of tracking which node is the root node of the entire tree. 

We could set the type to `"#document-fragment"` to stick more to the HTML specification, as `parse5` does.

File: src/tree/builder/TreeAdapter.js [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+	// Node source location.
+
+	/**
+	 * Sets the node's location as found in the source code.
+	 *
+	 * More often than not this is not the entire info
+	 * and misses the location of the end tag...
+	 *
+	 * We still need to add it, since `parse5` appends the end tag position
+	 * to this object somewhere during parsing.
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Since Natalia has made the issue I cannot apply suggestions in GitHub 😅.
I have made the change using the editor instead 🙂.

File: src/tree/builder/TreeAdapter.js [hunk 19]
```diff
@@ -0,0 +1,449 @@
...
+	 * Wraps a formatting element in a paragraph and adds the resulting paragraph to the given parent.
+	 *
+	 * @param {FormattingElement} formattingElement  The formatting element to wrap in a paragraph and add to the tree.
+	 * @param {Node} parent                          The parent element to add the new paragraph to.
+	 *
+	 * @returns {void}
+	 *
+	 * @private
+	 */
+	static _addOrphanedFormattingElement( formattingElement, parent ) {
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: Changed the order to make it consistent with the other functions!

File: src/tree/builder/TreeAdapter.js [hunk 20]
```diff
@@ -0,0 +1,449 @@
...
+	 * @param {Node|FormattingElement} element  The node to find the ancestor of.
+	 * @param {predicate} predicate	             The predicate to check the ancestors on.
+	 *
+	 * @returns {Node|null} The most recent ancestor that returns true on the given predicate.
+	 *
+	 * @private
+	 */
+	static _findAncestor( element, predicate ) {
+		let parent = element.parent;
+		while ( ! predicate( parent ) && parent.parent !== null ) {
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: I think this was a brainfart on my part.
I changed it so it checks if `parent` is `null` instead of `parent.parent`, which should make more sense.

I also added an extra inline comment explaining the workings of the loop. In my opinion adding an extra line of code would not make it more readable in this case.

File: src/tree/builder/TreeAdapter.js [hunk 21]
```diff
@@ -0,0 +1,449 @@
...
+		if ( parent instanceof StructuredNode ) {
+			// Wrap it in a paragraph, add it as formatting.
+			TreeAdapter._addOrphanedFormattingElement( formattingElement, parent );
+		} else {
+			/*
+			 Formatting elements can be nested, we want to add it to
+			 the most recent ancestor which is either a heading or paragraph.
+			 */
+			const ancestor = TreeAdapter._findAncestor( formattingElement, TreeAdapter._isLeafNode );
+			if ( TreeAdapter._isLeafNode( ancestor ) ) {
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: You're right, it now checks for `null` instead.

File: src/tree/builder/TreeAdapter.js (hunk 4 above)
Comment (hansjovis) [reply src/tree/builder/TreeAdapter.js]: I added two tests in `TreeAdapterSpec.js` that tests the `detachNode` functionality (both in the case of having a parent and not having one).

File: spec/tree/builder/buildTreeSpec.js (hunk 6 above)
Comment (hansjovis) [reply spec/tree/builder/buildTreeSpec.js]: Spelling is ~hart~ hard 😛 . Going to fix it!

File: src/tree/builder/buildTree.js [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+ *
+ * @param {string} html	The html-string that should be parsed.
+ * @returns {module:tree/structure.Node} The build tree.
+ *
+ * @memberOf module:tree/builder
+ */
+const buildTree = function( html ) {
+	const treeAdapter = new TreeAdapter();
+	let tree = parseFragment( html, { treeAdapter: treeAdapter, sourceCodeLocationInfo: true } );
+	// Cleanup takes < 2ms, even after parsing a big HTML file. (measured using `console.time`).
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/builder/buildTree.js]: Parsing of the full text tests takes 19ms on average (n=24). I added this as a comment!
(Most texts take < 1ms, 2ms is really an upper bound as far as I can see).

File: src/tree/structure/TextContainer.js [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+ *             endText: 24,	// "This text is very strong".length
+ *             attributes: {
+ *                 id: "elem-id"
+ *             }
+ *         }
+ *     ]
+ * }
+ * ```
+ *
+ * @memberOf module:tree/structure
```
[the remark this answers is not in the document]
Comment (hansjovis) [reply src/tree/structure/TextContainer.js]: Unfortunately JSDoc does not seem to play nice with this approach 🙁.
Having multiple `@module` comments in multiple files leads to multiple module definitions, instead of merging them into one.

See also: https://github.com/jsdoc3/jsdoc/issues/1002

**Comment by igorschoester** [comment]
CR 👍
Acceptance 👍  (tests look good and succeed, coverage 100%).

Very nice work @hansjovis ! 🎉
